// lucid-extension-sdk/core/unfurl/unfurlthumbnail.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeUnfurlThumbnail = exports.deserializeUnfurlThumbnail = exports.isValidUnfurlThumbnail = void 0;
const checks_1 = require("../checks");
const validators_1 = require("../validators/validators");
/** @ignore */
exports.isValidUnfurlThumbnail = (0, validators_1.objectValidator)({
    'url': checks_1.isString,
    'width': (0, validators_1.nullableOption)(checks_1.isNumber),
    'height': (0, validators_1.nullableOption)(checks_1.isNumber),
});
/** @ignore */
function deserializeUnfurlThumbnail(raw) {
    var _a, _b;
    return {
        url: raw['url'],
        width: (_a = raw['width']) !== null && _a !== void 0 ? _a : undefined,
        height: (_b = raw['height']) !== null && _b !== void 0 ? _b : undefined,
    };
}
exports.deserializeUnfurlThumbnail = deserializeUnfurlThumbnail;
/** @ignore */
function serializeUnfurlThumbnail(concrete) {
    return {
        'url': concrete.url,
        'width': concrete.width,
        'height': concrete.height,
    };
}
exports.serializeUnfurlThumbnail = serializeUnfurlThumbnail;
