// lucid-extension-sdk/core/unfurl/unfurldetails.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeUnfurlDetails = exports.deserializeUnfurlDetails = exports.isValidUnfurlDetails = void 0;
const checks_1 = require("../checks");
const validators_1 = require("../validators/validators");
const unfurliframe_1 = require("./unfurliframe");
const unfurlthumbnail_1 = require("./unfurlthumbnail");
/** @ignore */
exports.isValidUnfurlDetails = (0, validators_1.objectValidator)({
    'ProviderName': checks_1.isString,
    'ProviderFaviconUrl': checks_1.isString,
    'UnfurlTitle': (0, validators_1.option)(checks_1.isString),
    'Iframe': (0, validators_1.option)(unfurliframe_1.isValidUnfurlIframe),
    'Thumbnails': (0, validators_1.option)((0, validators_1.arrayValidator)(unfurlthumbnail_1.isValidUnfurlThumbnail)),
    'PreviewImageUrl': (0, validators_1.option)(checks_1.isString),
});
/** @ignore */
function deserializeUnfurlDetails(raw) {
    var _a;
    return {
        providerName: raw['ProviderName'],
        providerFaviconUrl: raw['ProviderFaviconUrl'],
        unfurlTitle: raw['UnfurlTitle'],
        iframe: raw['Iframe'] && (0, unfurliframe_1.deserializeUnfurlIframe)(raw['Iframe']),
        thumbnails: (_a = raw['Thumbnails']) === null || _a === void 0 ? void 0 : _a.map((raw) => (0, unfurlthumbnail_1.deserializeUnfurlThumbnail)(raw)),
        previewImageUrl: raw['PreviewImageUrl'],
    };
}
exports.deserializeUnfurlDetails = deserializeUnfurlDetails;
/** @ignore */
function serializeUnfurlDetails(concrete) {
    var _a;
    return {
        'ProviderName': concrete.providerName,
        'ProviderFaviconUrl': concrete.providerFaviconUrl,
        'UnfurlTitle': concrete.unfurlTitle,
        'Iframe': concrete.iframe && (0, unfurliframe_1.serializeUnfurlIframe)(concrete.iframe),
        'Thumbnails': (_a = concrete.thumbnails) === null || _a === void 0 ? void 0 : _a.map((raw) => (0, unfurlthumbnail_1.serializeUnfurlThumbnail)(raw)),
        'PreviewImageUrl': concrete.previewImageUrl,
    };
}
exports.serializeUnfurlDetails = serializeUnfurlDetails;
