// lucid-extension-sdk/core/properties/fillcolor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeSimpleFill = exports.SimpleImageFillPosition = void 0;
const checks_1 = require("../checks");
/**
 * Different ways of laying out the image in the background of the block.
 */
var SimpleImageFillPosition;
(function (SimpleImageFillPosition) {
    /** The image retains its original aspect ratio, and is as large as possible while still fitting in the block */
    SimpleImageFillPosition["Fit"] = "fit";
    /** The image retains its original aspect ratio, and is as small as possible while still completely covering the block */
    SimpleImageFillPosition["Fill"] = "fill";
    /** The image is stretched to the bounding box of the block */
    SimpleImageFillPosition["Stretch"] = "stretch";
    /** The image is shown in its original size if possible */
    SimpleImageFillPosition["Original"] = "original";
    /** The image is tiled in its original size vertically and horizontally to fully cover the block */
    SimpleImageFillPosition["Tile"] = "tile";
    //TODO: 'mask' but mask requires additional settings
})(SimpleImageFillPosition || (exports.SimpleImageFillPosition = SimpleImageFillPosition = {}));
/** @ignore */
function serializeSimpleFill(fill) {
    if ((0, checks_1.isString)(fill)) {
        return fill;
    }
    else {
        return {
            'pos': fill.position,
            'url': fill.url,
            'polys': null,
        };
    }
}
exports.serializeSimpleFill = serializeSimpleFill;
