// lucid-extension-sdk/core/properties/strokestyle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertStrokeStyleToBorder = exports.isStrokeStyle = exports.StrokeStyle = void 0;
var StrokeStyle;
(function (StrokeStyle) {
    StrokeStyle["Solid"] = "solid";
    StrokeStyle["Dashed"] = "dashed";
    StrokeStyle["LongDash"] = "longdash";
    StrokeStyle["DashDot"] = "dashdot";
    StrokeStyle["DashDotDot"] = "dashdotdot";
    StrokeStyle["DashLongDash"] = "dashlongdash";
    StrokeStyle["Dashed32"] = "dashed32";
    StrokeStyle["Dashed44"] = "dashed44";
    StrokeStyle["Dashed24"] = "dashed24";
    StrokeStyle["Dotted"] = "dotted";
    StrokeStyle["DotDotDot"] = "dotdotdot";
    StrokeStyle["StripeThinThin"] = "stripethinthin";
    StrokeStyle["StripeThickThick"] = "stripethickthick";
    StrokeStyle["StripeThickThin"] = "stripethickthin";
    StrokeStyle["StripeThinThick"] = "stripethinthick";
    StrokeStyle["StripeThickThinThick"] = "stripethickthinthick";
    StrokeStyle["StripeThinThickThin"] = "stripethinthickthin";
    StrokeStyle["StripeTriple"] = "stripetriple";
})(StrokeStyle || (exports.StrokeStyle = StrokeStyle = {}));
const strokeStyleSet = new Set(Object.values(StrokeStyle));
function isStrokeStyle(value) {
    return strokeStyleSet.has(value);
}
exports.isStrokeStyle = isStrokeStyle;
function convertStrokeStyleToBorder(strokeStyle) {
    if (strokeStyle.startsWith('dash') || strokeStyle == StrokeStyle.LongDash) {
        return 'dashed';
    }
    else if (strokeStyle.startsWith('dot')) {
        return 'dotted';
    }
    else {
        return 'solid';
    }
}
exports.convertStrokeStyleToBorder = convertStrokeStyleToBorder;
