// lucid-extension-sdk/core/offsettype.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isLinearOffsetType = exports.OffsetType = void 0;
/**
 * When adjusting the size or location of items on a page, these are the specific operations that are available.
 */
var OffsetType;
(function (OffsetType) {
    /**
     * Resize from the northwest (upper-left) corner of the selection, anchoring the lower-right corner in place.
     */
    OffsetType[OffsetType["NW"] = 0] = "NW";
    /**
     * Resize from the northeast (upper-right) corner of the selection, anchoring the lower-left corner in place.
     */
    OffsetType[OffsetType["NE"] = 1] = "NE";
    /**
     * Resize from the southeast (lower-right) corner of the selection, anchoring the upper-left corner in place.
     */
    OffsetType[OffsetType["SE"] = 2] = "SE";
    /**
     * Resize from the southwest (lower-left) corner of the selection, anchoring the upper-right corner in place.
     */
    OffsetType[OffsetType["SW"] = 3] = "SW";
    /**
     * Move the selection without resizing.
     */
    OffsetType[OffsetType["MOVE"] = 4] = "MOVE";
    OffsetType[OffsetType["ROTATE"] = 5] = "ROTATE";
    OffsetType[OffsetType["SCALE"] = 6] = "SCALE";
    OffsetType[OffsetType["CUSTOM"] = 7] = "CUSTOM";
    /**
     * Resize from the north (upper) side of the selection, stretching vertically.
     */
    OffsetType[OffsetType["N"] = 8] = "N";
    /**
     * Resize from the east (right) side of the selection, stretching horizontally.
     */
    OffsetType[OffsetType["E"] = 9] = "E";
    /**
     * Resize from the south (lower) side of the selection, stretching vertically.
     */
    OffsetType[OffsetType["S"] = 10] = "S";
    /**
     * Resize from the west (left) side of the selection, stretching horizontally.
     */
    OffsetType[OffsetType["W"] = 11] = "W";
})(OffsetType || (exports.OffsetType = OffsetType = {}));
function isLinearOffsetType(offsetType) {
    return (offsetType === OffsetType.NW ||
        offsetType === OffsetType.NE ||
        offsetType === OffsetType.SE ||
        offsetType === OffsetType.SW ||
        offsetType === OffsetType.MOVE ||
        offsetType === OffsetType.N ||
        offsetType === OffsetType.E ||
        offsetType === OffsetType.S ||
        offsetType === OffsetType.W);
}
exports.isLinearOffsetType = isLinearOffsetType;
