// lucid-extension-sdk/core/xhr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isXHRResponse = exports.isBinaryXHRResponse = exports.isTextXHRResponse = void 0;
const checks_1 = require("./checks");
const validators_1 = require("./validators/validators");
const isBaseXHRResponse = (0, validators_1.objectValidator)({
    url: checks_1.isString,
    status: checks_1.isNumber,
    headers: (0, validators_1.objectOfValidator)(checks_1.isString),
    timeout: (0, validators_1.option)(checks_1.isBoolean),
});
exports.isTextXHRResponse = (0, validators_1.both)(isBaseXHRResponse, (0, validators_1.objectValidator)({
    responseFormat: (0, checks_1.isLiteral)('utf8'),
    responseText: checks_1.isString,
}));
exports.isBinaryXHRResponse = (0, validators_1.both)(isBaseXHRResponse, (0, validators_1.objectValidator)({
    responseFormat: (0, checks_1.isLiteral)('binary'),
    responseData: (0, checks_1.isInstanceOf)(Uint8Array),
}));
exports.isXHRResponse = (0, validators_1.either)(exports.isTextXHRResponse, exports.isBinaryXHRResponse);
