// lucid-extension-sdk/core/properties/datagraphic/datagraphicindexes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.areDataGraphicImagesEqual = exports.areDataGraphicIconIndexesEqual = exports.DataGraphicIconSets = void 0;
var DataGraphicIconSets;
(function (DataGraphicIconSets) {
    // Custom icons are a user-defined set of uploaded icons, not a predefined set of icons
    DataGraphicIconSets[DataGraphicIconSets["CUSTOM_ICONS"] = -1] = "CUSTOM_ICONS";
    DataGraphicIconSets[DataGraphicIconSets["STATUS_ICONS"] = 0] = "STATUS_ICONS";
    DataGraphicIconSets[DataGraphicIconSets["STATUS_CIRCLES"] = 1] = "STATUS_CIRCLES";
    DataGraphicIconSets[DataGraphicIconSets["STOPLIGHTS"] = 2] = "STOPLIGHTS";
    DataGraphicIconSets[DataGraphicIconSets["TOGGLES"] = 3] = "TOGGLES";
    DataGraphicIconSets[DataGraphicIconSets["COMPLETION_PIES"] = 4] = "COMPLETION_PIES";
    DataGraphicIconSets[DataGraphicIconSets["TRENDING_ARROWS"] = 5] = "TRENDING_ARROWS";
    DataGraphicIconSets[DataGraphicIconSets["STATUS_CHECK_BOXES"] = 6] = "STATUS_CHECK_BOXES";
    DataGraphicIconSets[DataGraphicIconSets["TRENDING_ARROWS_FLIPPED"] = 7] = "TRENDING_ARROWS_FLIPPED";
})(DataGraphicIconSets || (exports.DataGraphicIconSets = DataGraphicIconSets = {}));
function areDataGraphicIconIndexesEqual(a, b) {
    return a.set == b.set && a.index == b.index && areDataGraphicImagesEqual(a.iconImage, b.iconImage);
}
exports.areDataGraphicIconIndexesEqual = areDataGraphicIconIndexesEqual;
function areDataGraphicImagesEqual(a, b) {
    if (a == undefined || b == undefined) {
        return a == b;
    }
    return a.w == b.w && a.h == b.h && a.url == b.url;
}
exports.areDataGraphicImagesEqual = areDataGraphicImagesEqual;
