// lucid-extension-sdk/core/properties/datagraphic/staticdatagraphicsettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSerializedSimpleStaticDataGraphicSettings = exports.deserializeSimpleStaticDataGraphicSettings = exports.serializeSimpleStaticDataGraphicSettings = void 0;
const checks_1 = require("../../checks");
const badgeposition_1 = require("./badgeposition");
function serializeSimpleStaticDataGraphicSettings(settings) {
    return {
        'icon': settings.icon,
        'pos': (0, badgeposition_1.serializeBadgeEnumPosition)(settings.position),
        'color': settings.color,
    };
}
exports.serializeSimpleStaticDataGraphicSettings = serializeSimpleStaticDataGraphicSettings;
function deserializeSimpleStaticDataGraphicSettings(settings) {
    return {
        icon: settings['icon'],
        position: (0, badgeposition_1.deserializeBadgeEnumPosition)(settings['pos']),
        color: settings['color'],
    };
}
exports.deserializeSimpleStaticDataGraphicSettings = deserializeSimpleStaticDataGraphicSettings;
function isSerializedSimpleStaticDataGraphicSettings(settings) {
    return ((0, checks_1.isObjectUnsafe)(settings) &&
        (0, checks_1.isString)(settings['color']) &&
        (0, checks_1.isObject)(settings['icon']) &&
        (0, badgeposition_1.isValidSerializedBadgeEnumPosition)(settings['pos']));
}
exports.isSerializedSimpleStaticDataGraphicSettings = isSerializedSimpleStaticDataGraphicSettings;
