// lucid-extension-sdk/document/documentelement/documentelementtype.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentElementType = void 0;
/**
 * An enumeration of document element types. A Document element represents a collection of data that is stored on a document.
 */
var DocumentElementType;
(function (DocumentElementType) {
    DocumentElementType["DataTransformation"] = "DerivedStructure";
    DocumentElementType["Path"] = "Path";
    DocumentElementType["Tag"] = "Tag";
    DocumentElementType["CommentThreadMetadata"] = "ThreadMetadata";
    DocumentElementType["FontStylePreset"] = "FontStylePreset";
    DocumentElementType["IntraDocumentMutex"] = "Mutex";
    DocumentElementType["Rule"] = "Rule";
    DocumentElementType["DocumentFormula"] = "Formula";
    DocumentElementType["TaskCardFieldsConfig"] = "TaskCardFieldsConfig";
    DocumentElementType["GeneratorView"] = "GeneratorView";
    DocumentElementType["ShapeStylePreset"] = "ShapeStylePreset";
    DocumentElementType["ShapeSpecificDefaultProperties"] = "ShapeSpecificDefaultProperties";
})(DocumentElementType || (exports.DocumentElementType = DocumentElementType = {}));
