// lucid-extension-sdk/document/linedefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPositionEndpointDefinition = exports.isLineEndpointDefinition = exports.isBlockEndpointDefinition = void 0;
const checks_1 = require("../core/checks");
const validators_1 = require("../core/validators/validators");
const blockproxy_1 = require("./blockproxy");
const lineproxy_1 = require("./lineproxy");
/**
 * Checks if the endpoint is a BlockEndpointDefinition.
 */
function isBlockEndpointDefinition(endpoint) {
    return (0, validators_1.objectValidator)({
        connection: (0, checks_1.isInstanceOf)(blockproxy_1.BlockProxy),
        linkX: checks_1.isNumber,
        linkY: checks_1.isNumber,
        inside: (0, validators_1.option)(checks_1.isBoolean),
        autoLink: (0, validators_1.option)(checks_1.isBoolean),
        padding: (0, validators_1.option)(checks_1.isNumber),
    })(endpoint);
}
exports.isBlockEndpointDefinition = isBlockEndpointDefinition;
/**
 * Checks if the endpoint is a LineEndpointDefinition.
 */
function isLineEndpointDefinition(endpoint) {
    return (0, validators_1.objectValidator)({
        connection: (0, checks_1.isInstanceOf)(lineproxy_1.LineProxy),
        position: checks_1.isNumber,
    })(endpoint);
}
exports.isLineEndpointDefinition = isLineEndpointDefinition;
/**
 * Checks if the endpoint is a PositionEndpointDefinition.
 */
function isPositionEndpointDefinition(endpoint) {
    return (0, validators_1.objectValidator)({
        x: checks_1.isNumber,
        y: checks_1.isNumber,
    })(endpoint);
}
exports.isPositionEndpointDefinition = isPositionEndpointDefinition;
