// lucid-extension-sdk/document/linetextareapositioning.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeLineTextAreaPositioning = exports.deserializeLineTextAreaPositioning = exports.isSerializedLineTextAreaPositioning = void 0;
const checks_1 = require("../core/checks");
/** @ignore */
function isSerializedLineTextAreaPositioning(raw) {
    return ((0, checks_1.isObjectUnsafe)(raw) &&
        (0, checks_1.isNumber)(raw['Location']) &&
        (0, checks_1.isNumber)(raw['Side']) &&
        ((0, checks_1.isBoolean)(raw['AllowOverBlock']) || (0, checks_1.isUndefined)(raw['AllowOverBlock'])));
}
exports.isSerializedLineTextAreaPositioning = isSerializedLineTextAreaPositioning;
/** @ignore */
function deserializeLineTextAreaPositioning(raw) {
    return {
        location: raw['Location'],
        side: raw['Side'],
        allowOverBlock: raw['AllowOverBlock'],
    };
}
exports.deserializeLineTextAreaPositioning = deserializeLineTextAreaPositioning;
/** @ignore */
function serializeLineTextAreaPositioning(data) {
    return {
        'Location': data.location,
        'Side': data.side,
        'AllowOverBlock': data.allowOverBlock,
    };
}
exports.serializeLineTextAreaPositioning = serializeLineTextAreaPositioning;
