// lucid-extension-sdk/core/cardintegration/cardintegrationconfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeCardIntegrationConfig = exports.serializeCardIntegrationConfig = void 0;
const cardfielddisplaysettings_1 = require("./cardfielddisplaysettings");
/** @ignore */
function serializeCardIntegrationConfig(config) {
    return {
        'cc': {
            'f': config.cardConfig.fieldNames,
            'fd': config.cardConfig.fieldDisplaySettings
                ? [...config.cardConfig.fieldDisplaySettings].map(([key, settings]) => [
                    key,
                    (0, cardfielddisplaysettings_1.serializeLucidCardFieldDisplaySettings)(settings),
                ])
                : undefined,
            'fs': config.cardConfig.fieldStyles ? [...config.cardConfig.fieldStyles] : undefined,
        },
        'cdpc': {
            'f': config.cardDetailsPanelConfig.fields.map((field) => {
                return {
                    'n': field.name,
                    'l': field.locked,
                };
            }),
        },
    };
}
exports.serializeCardIntegrationConfig = serializeCardIntegrationConfig;
/** @ignore */
function deserializeCardIntegrationConfig(raw) {
    return {
        cardConfig: {
            fieldNames: raw['cc']['f'],
            fieldStyles: raw['cc']['fs'] ? new Map(raw['cc']['fs']) : undefined,
            fieldDisplaySettings: raw['cc']['fd']
                ? new Map(raw['cc']['fd'].map(([key, settings]) => [
                    key,
                    (0, cardfielddisplaysettings_1.deserializeLucidCardFieldDisplaySettings)(settings),
                ]))
                : undefined,
        },
        cardDetailsPanelConfig: {
            fields: raw['cdpc']['f'].map((field) => {
                return {
                    name: field['n'],
                    locked: field['l'],
                };
            }),
        },
    };
}
exports.deserializeCardIntegrationConfig = deserializeCardIntegrationConfig;
