// lucid-extension-sdk/document/blockclasses/erdblockproxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERDBlockProxy = exports.ERDFieldProxy = void 0;
const blockproxy_1 = require("../blockproxy");
class ERDFieldProxy {
    constructor(block, index) {
        this.block = block;
        this.index = index;
    }
    getName() {
        return this.block.properties.get('Field' + this.index);
    }
    getType() {
        var _a;
        return (_a = this.block.properties.get('Type' + this.index)) !== null && _a !== void 0 ? _a : '';
    }
    getKey() {
        var _a;
        return (_a = this.block.properties.get('Key' + this.index)) !== null && _a !== void 0 ? _a : '';
    }
}
exports.ERDFieldProxy = ERDFieldProxy;
class ERDBlockProxy extends blockproxy_1.BlockProxy {
    getName() {
        return this.properties.get('Name');
    }
    getFieldCount() {
        return this.properties.get('Fields');
    }
    getFields() {
        const fields = [];
        const fieldCount = this.getFieldCount();
        for (let i = 1; i <= fieldCount; i++) {
            //These are 1-indexed in the property store
            fields.push(new ERDFieldProxy(this, i));
        }
        return fields;
    }
}
exports.ERDBlockProxy = ERDBlockProxy;
ERDBlockProxy.classNameRegex = /^ERDEntityBlock(2|3|4)?$/;
