// lucid-extension-sdk/document/text/textstyle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPartialTextStyle = exports.TextMarkupNames = void 0;
const checks_1 = require("../../core/checks");
const validators_1 = require("../../core/validators/validators");
/**
 * Text styles that can be read and written with ItemProxy.textStyles.
 */
var TextMarkupNames;
(function (TextMarkupNames) {
    TextMarkupNames["Family"] = "font";
    TextMarkupNames["Bold"] = "bold";
    TextMarkupNames["Italic"] = "italic";
    TextMarkupNames["Underline"] = "underline";
    TextMarkupNames["Size"] = "size";
    TextMarkupNames["Color"] = "color";
    TextMarkupNames["HAlign"] = "align";
})(TextMarkupNames || (exports.TextMarkupNames = TextMarkupNames = {}));
exports.isPartialTextStyle = (0, validators_1.partialObjectValidator)({
    [TextMarkupNames.Family]: checks_1.isString,
    [TextMarkupNames.Bold]: checks_1.isBoolean,
    [TextMarkupNames.Italic]: checks_1.isBoolean,
    [TextMarkupNames.Underline]: checks_1.isBoolean,
    [TextMarkupNames.Size]: checks_1.isNumber,
    [TextMarkupNames.Color]: checks_1.isString,
    [TextMarkupNames.HAlign]: (value) => value === 'left' || value === 'right' || value === 'center',
});
